### ISSUE TITLE
[feature request] trackingDataCalibrationScan.py Calibration Scans w/full data path.

### ISSUE THREAD

**Issue by bdorney** [cms-gem-daq-project/cmsgemos#284]
## Brief summary of issue

Provide a tool to take calibration scans using VFAT3 calibration module and full data path to allow for multiple AMCs/links.

Should target branch `generic-amc-RPC-v3-short-term`.

Going to need:
- https://github.com/cms-gem-daq-project/ctp7_modules/issues/129
- https://github.com/cms-gem-daq-project/xhal/issues/118

### Types of issue

- [ ] Bug report (report an issue with the code)
- [x] Feature request (request for change which adds functionality)

### Supporting Change to `HwVFAT`
You'll need to use the function you implemented in the two issues above and add a wrapper around it in the `HwVFAT` class.  Inside `HwVFAT::__init__` you'll need to add something like:

```python
self.confCalPulse = self.parentOH.parentAMC.lib.confCalPulse
self.confCalPulse.argTypes = [ c_uint, c_uint, c_uint, c_bool, c_bool, c_uint ]
self.confCalPulse.restype = c_uint 
```

Then you'll need to add some class method to `HwVFAT` e.g.:

```python
def configureCalPulseAllVFATs(self, ch, toggleOn, mask=0x0, currentPulse=False,  calScaleFactor=0x0):
    """
    Function documentation goes here
    """

    # some check on the validity of the input arguments may be required

    return self.confCalPulse( args go here)
```

### Calibration Scans w/full data path: `trackingDataCalibrationScan.py`
This will allow you to use the VFAT3 calibration module and the AMC13 to take calibration scans with the full data path.  It's understood this is used on the following FM in the 904 RCMS instance under QC8 folder:

- `LocalReadout_noLEMO`
- `uFEDKit_noLEMO`

The following would be the process flow:

1. Create an `AMC13` class object and stop triggers:
```python
    import amc13, os
    connection_file = "%s/connections.xml"%(os.getenv("GEM_ADDRESS_TABLE_PATH"))
    amc13base  = "gem.shelf%02d.amc13"%(args.shelf)
    amc13board = amc13.AMC13(connection_file,"%s.T1"%(amc13base),"%s.T2"%(amc13base))
    amc13board.enableLocalL1A(False)
    amc13board.stopContinuousL1A()
```
2. Declare the following dictionaries:
    - `dict_vfatBoards` whose keys are AMC slot numbers and the values are instances of `HwVFAT`, each `HwVFAT` object declared here should have  some dummy link value but the correct slot number.  You will update the link number on the fly.
    - `dict_ohMasks` whose keys are AMC slot number, values are a 12 bit number specifying which OH's to use (0x1 means use this OH)
    - `dict_vfatMask` whose keys are AMC slot number, value is a ctypes array of size 12 whose elements are 24 bit numbers storing VFAT masks (0x1 means skip this VFAT).
3. Loop over the members `dict_vfatBoards`and discover which OH's are on this AMC and each of their `vfatmask`s using:
```python
dict_ohMasks[slot] = dict_vfatBoards[slot].parentOH.parentAMC.getOHMask()
dict_vfatMasks[slot] = dict_vfatBoards[slot].parentOH.parentAMC. getMultiLinkVFATMask(dict_ohMasks[slot])
```
4. Inside the loop (mentioned in step 3) over the members `dict_vfatBoards` for each unmasked OH on this AMC do the following:
```python
        for ohN in range(dict_vfatBoards[slot].parentOH.parentAMC.nOHs):
            if (dict_ohMasks[slot] >> ohN) & 0x0: continue
            
            dict_vfatBoards[slot].parentOH.link = ohN 
           
            # Stop Cal Pulses
            dict_vfatBoards[slot].stopCalPulses(dict_vfatMasks[slot][ohN]) 

            # Set Parameters
            dict_vfatBoards[slot].setVFATLatencyAll(mask=dict_vfatMasks[slot][ohN], lat=args.latency)
            dict_vfatBoards[slot].setVFATMSPLAll(mask=dict_vfatMasks[slot][ohN], mspl=args.pulseStretch)
            dict_vfatBoards[slot].setVFATCalPhaseAll(mask=dict_vfatMasks[slot][ohN], phase =args.calPhase)
            if args.runType == 0x2:
                dict_vfatBoards[slot].setVFATCalHeightAll(mask=dict_vfatMasks[slot][ohN], currentPulse=False)
```
5. Configure AMC13 Trigger mode
    - I'll Update this later but for now use the following block:
```python
    # Config BGO generator to send a Cal Pulse before the L1A
    cmd_calPulse = 0x14
    amc13board.configureBGOShort(0, cmd_calPulse, 450, 0, True)
    
    # Configure locally generated triggers for one L1A per orbit @ BX=500
    amc13board.configureLocalL1A(True, 0, 1, 1, 0)
```
6. Scan over the user requested range [`dacMin`, `dacMax`] in the user specified `dacStep` (this implies the below is inside a for loop):
    - Stop triggers from AMC13, see last two lines of step 1 above
    - Loop over all slots in `dict_vfatBoards`, for each value in `dict_vfatBoards` do the following
        - Write `0xf` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_TYPE`
        - Write `0x0` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_PARAMS`
        - Loop over each unmasked OH within this slot and set the current DAC value to the scan register (see correspondence between user supplied Run Type and scan register in Table 1 below).
    - If `RunType == 0x3` you're going to send a user supplied number of triggers (e.g args.nevt), first call `amc13board.enableLocalL1A(True)`
        - Write `args.runType` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_TYPE`
        - Write the current `dacValue` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_PARAMS`
        - Loop over all unmasked OH's within this slot and set the VFATs into run mode with `HwVFAT:: setRunModeAll`
        - Then inside a for loop over args.nevt call `amc13board.sendL1ABurst()`
        - Loop over all unmasked OH's within this slot and take the VFATs out of run mode with `HwVFAT:: setRunModeAll`
        - Write `0xf` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_TYPE`
        - Write `0x0` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_PARAMS`
        - You now move to the next `dacValue`
    - If `RunType == 0x2 || RunType == 0x2 ` you're going to loop over channels `args.chanMin` to `args.chanMax+1`
        - Loop over all unmasked OH's within this slot and:
            - Configure the calibration pulse for this channel using `HwVFAT::configureCalPulseAllVFATs`
            - Set the VFATs into run mode with `HwVFAT:: setRunModeAll`
        - Write `args.runType` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_TYPE`
        - Write the current `dacValue` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_PARAMS`
        - Then inside a for loop over args.nevt call `amc13board.sendL1ABurst()`
        - Loop over all unmasked OH's within this slot and:
            - Stop the calibration pulse for this channel using `HwVFAT::configureCalPulseAllVFATs`
            - Take the VFATs out of run mode with `HwVFAT:: setRunModeAll`
        - Write `0xf` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_TYPE`
        - Write `0x0` into `GEM_AMC.DAQ.EXT_CONTROL.RUN_PARAMS`
        - You now move to the next `dacValue`

This should then enable a generic tool to conduct tracking data scans using the calibration module of the VFAT3 with the AMC13.

The user then needs to be able to supply the following:

- uTCA shelf number (positional argument),
- List of AMC's to be considered, should be a string (positional argument), e.g. expect something like `2,4,6` or `2,4-9` where the second should be understood as 4 through 9 (you'll need to develop some function to parse this)
 - Run Type following from https://github.com/cms-gem-daq-project/cmsgemos/issues/230 so acceptable values are {2,3,4} (positional argument),
 - `CFG_LATENCY` (optional argument, e.g. args.latency above, default to 33 for now),
 - `CFG_CAL_PHI` (optional argument, default to 0)
 - `dacMin` starting DAC value, optional, default to 0
 - `dacMax` ending DAC value, optional, default to 255 (but know that `CFG_LATENCY` can go to 1023)
 - `dacStep` step size for scanning from `dacMin` to `dacMax`, optional, default to 1.
 - `chanMin` starting VFAT channel, optional, default to 0
 - `chanMax` ending VFAT channel, optional, default to 127

Table 1. Correspondence between RunType, DacName and how to set these with methods from `HwVFAT`.

| RunType | Dac Name | Function To Change DAQ Value | Note |
| :--: | :--: | :--: |  :-- |
| 0x2 | `CFG_LATENCY` | `HwVFAT:: setVFATLatencyAll` | |
| 0x3 | `CFG_THR_ARM_DAC` | `HwVFAT::setVFATThresholdAll ` | |
| 0x4 | `CFG_CAL_DAC` | `HwVFAT::setVFATCalHeightAll ` | always use `currentPulse=False` |

Further details on the `AMC13` python API can be found:
http://cactus.web.cern.ch/cactus/release/amc13/1.1/api/html/d5/df6/classamc13_1_1_a_m_c13.html

## Current Behavior

No tool exists to use the full data path with tracking data scans.

## Context

Need a stop gap measure between when calibration suite in xdaq is ready and now for QC8.

**Comment by bdorney** [comment]
Okay I think I haven't forgotten anything.  But before @ram1123 get's started it would be good if @mexanick or @jsturdy could review this.

Step 5 above might be updated when we here back from BU regarding how to configure AMC13 to have a fixed delay between L1A and BGO; @evka85 is also working on a FW functionality which would enable the CTP7 to delay an L1A from the AMC13 and send a calpulse then the L1A.  This would then remove the limitation of only being able to send 1-4 calpulses per orbit.  Once this is available we will replace step 5 with this.

**Comment by bdorney** [comment]
In the meantime @ram1123 the two issues in `xhal` and `ctp7_modules` are straight forward enough that you could get started already on them.

**Comment by bdorney** [comment]
I will have a prototype to test hopefully by the end of the day or tomorrow.

**Comment by bdorney** [comment]
Following the discussion in yesterday's SW dev meeting we decided that during scurve scans all channels should be masked *except* the channel being pulsed to prevent any potential cross-talk issues or stray hits (e.g. from particles/noise/other sources)

**Comment by bdorney** [comment]
Closed by #302
